src/search/BracketsSearchCursor.js

```diff
@@ -0,0 +1,823 @@
+     * @param {String} text The string to index
+     * @param {String} lineSeparator The ending character that splits lines
+     */
+    function _createLineCharacterCountIndex(text, lineSeparator) {
+        var lineNumber;
+        // splitting is actually faster than using doc.getLine()
+        var lines = text.split(lineSeparator);
+        var lineSeparatorLength = lineSeparator.length;
+        var lineCharacterCountIndex = new Uint32Array(lines.length);
+        var lineCount = lines.length;
+        var totalCharacterCount = 0;
+        for (lineNumber = 0; lineNumber < lineCount; lineNumber++) {
+            totalCharacterCount += lines[lineNumber].length + lineSeparatorLength;
+            lineCharacterCountIndex[lineNumber] = totalCharacterCount;
+        }
+        return lineCharacterCountIndex;
+    }
+
+    /**
+     * Creates the document index and store in our map
+     * @param {CodeMirror.Doc} doc The codemirror document
+     */
+    function _indexDocument(doc) {
+        var docText = doc.getValue();
+        var docLineIndex = _createLineCharacterCountIndex(docText, doc.lineSeparator());
+        _documentMap.set(doc, {text: docText, index: docLineIndex, generation: doc.history.generation});
+    }
+
+    /**
+     * Gets the document index
+     * @param {CodeMirror.Doc} doc The codemirror document
+     * @returns {Array<number>} see '_createLineCharacterCountIndex' for contents of Array
+     */
+    function _getDocumentIndex(doc) {
+        return _documentMap.get(doc).index;
+    }
+
+    /**
+     * Gets the document text
+     * @param {CodeMirror.Doc} doc The codemirror document
+     * @returns {String} document text content
+     */
+    function _getDocumentText(doc) {
+        return _documentMap.get(doc).text;
+    }
+
+
+    /**
+     * Converts plain text query into regular expression
+     * If already regular expression, then just set the flags as appropriate
+     * @param {Object}  stringOrRegex A string or regular expression
+     * @param {boolean} ignoreCase True to ignore case for searchers
+     */
+    function _convertToRegularExpression(stringOrRegex, ignoreCase) {
+        if (typeof stringOrRegex === "string") {
+            return new RegExp(StringUtils.regexEscape(stringOrRegex), ignoreCase ? "igm" : "gm");
+        } else {
+            return new RegExp(stringOrRegex.source, ignoreCase ? "igm" : "gm");
+        }
+    }
+
+    /**
+     * Finds the line number for the given index
+     * @private
+     * @param {Array<number>} lineCharacterCountIndexArray See '_createLineCharacterCountIndex'
+     * @param {number}        startSearchingWithLine       Line number to start search
+     * @param {number}        indexWithinDoc               The index of the character offset from the start of the document
+     * @returns {number}      The line number for the given index.
+     */
+    function _lineFromIndex(lineCharacterCountIndexArray, startSearchingWithLine, indexWithinDoc) {
```

Comment:
`lineStart` could be a more concise name than `startSearchingWithLine`, also `charStart` instead of `indexWithinDoc`.